### DISCUSSION TITLE
Cruise control: Getting error message - Error processing POST request '/rebalance' due to: 'com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException: java.lang.IllegalArgumentException: Missing hard goals

### DISCUSSION THREAD

**Discussion by Gulnaz2021** [strimzi/strimzi-kafka-operator#4376] [General]
Hi,
With Strimzi operator version 20, I've applied the kafka-rebalance file to my namespace with kafka cluster. And upon describing the kafkarebalance using the command - kubectl describe kafkarebalance my-rebalance -n myproject, getting the below mentioned error:
    Message:               Error for request: my-cluster-cruise-control.kafka-cruise.svc:9090/kafkacruisecontrol/rebalance?json=true&dryrun=true&verbose=false&skip_hard_goal_check=false&goals=CpuCapacityGoal%2CNetworkInboundCapacityGoal%2CDiskCapacityGoal%2CRackAwareGoal%2CNetworkOutboundCapacityGoal. Server returned: Error processing POST request '/rebalance' due to: 'com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException: java.lang.IllegalArgumentException: Missing hard goals [NetworkInboundCapacityGoal, DiskCapacityGoal, RackAwareGoal, NetworkOutboundCapacityGoal, CpuCapacityGoal, ReplicaCapacityGoal] in the provided goals: [CpuCapacityGoal, NetworkInboundCapacityGoal, DiskCapacityGoal, RackAwareGoal, NetworkOutboundCapacityGoal]. Add skip_hard_goal_check=true parameter to ignore this sanity check.'.

Kafka-rebalance yaml file configuration:
  metadata:
    name: my-rebalance
    labels:
      strimzi.io/cluster: my-cluster
  spec:
    goals:
      - CpuCapacityGoal
      - NetworkInboundCapacityGoal
      - DiskCapacityGoal
      - RackAwareGoal
      - NetworkOutboundCapacityGoal
  
And for kafka yaml file, mentioned the cruise control as below.
   cruiseControl: {}

Can I please get some help in this regard. 
Thanks.

**Comment by tomncooper** [comment]
You have described a set of custom goals in your `KafkaRebalance` resource but that list does not contain all the configured (default in this case) hard goals, hence the `java.lang.IllegalArgumentException: Missing hard goals` error. 

As per [the Strimzi Cruise Control documentation](https://strimzi.io/docs/operators/latest/using.html#cruise-control-concepts-str) you must include all configured hard goal in your custom goals. 

You can either add the missing hard goals to the goals list in your `KafkaRebalance` resource or reconfigure (and restart) the CC server with a different set of hard goals (although be aware that will effect *all* rebalances).

**Comment by Gulnaz2021** [comment]
Thankyou! That helps!
